# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/webrender_bindings/DCLayerTree.cpp
# Commit: 31587f0a98cd
# Full Hash: 31587f0a98cd6b7a445c59ccd18c2207259dafd2
# Author: Kelsey Gilbert <kelsey.gilbert@mozilla.com>
# Date: 2023-03-14 09:41:39
# Regressor Bug: 1799258
# File Overlap Count: 1
# Description:
#   Bug 1799258 - Share all-of-dcomp.h preamble, and deal with outdated mingw dcomp.h. r=gfx-reviewers,sotaro
#   
#   Mingw's dcomp.h is not the official one, but rather a by-hand
#   reproduction. While this newly-updated version has e.g.
#   IDCompositionFilterEffect, it is still missing e.g.
# ==============================================================================

diff -r ab2b044c704c -r 31587f0a98cd gfx/webrender_bindings/DCLayerTree.cpp
--- a/gfx/webrender_bindings/DCLayerTree.cpp	Mon Mar 13 21:04:12 2023 +0000
+++ b/gfx/webrender_bindings/DCLayerTree.cpp	Mon Mar 13 21:04:12 2023 +0000
@@ -8,33 +8,9 @@
 
 // -
 
-#if (_WIN32_WINNT < _WIN32_WINNT_WIN10)
-
-#  define XSTR(x) STR(x)
-#  define STR(x) #x
-// clang-format off
-
-#  pragma message "IDCompositionFilterEffect in dcomp.h requires _WIN32_WINNT >= _WIN32_WINNT_WIN10."
-// Pedantically, it actually requires _WIN32_WINNT_WINTHRESHOLD, but that's the
-// same as _WIN32_WINNT_WIN10.
-
-#  pragma message "Forcing NTDDI_VERSION " XSTR(NTDDI_VERSION) " -> " XSTR(NTDDI_WIN10)
-#  undef NTDDI_VERSION
-#  define NTDDI_VERSION NTDDI_WIN10
-
-#  pragma message "Forcing _WIN32_WINNT " XSTR(_WIN32_WINNT) " -> " XSTR(_WIN32_WINNT_WIN10)
-#  undef _WIN32_WINNT
-#  define _WIN32_WINNT _WIN32_WINNT_WIN10
-
-// clang-format on
-#  undef STR
-#  undef XSTR
-
-#endif
-
+#include "mozilla/gfx/AllOfDcomp.h"
 #include <d3d11.h>
 #include <d3d11_1.h>
-#include <dcomp.h>
 #include <dxgi1_2.h>
 
 // -
@@ -57,6 +33,16 @@
 #include "nsPrintfCString.h"
 #include "WinUtils.h"
 
+// -
+
+#if defined(__MINGW32__)  // 64 defines both 32 and 64
+// We need to fake some things, while we wait on updates to mingw's dcomp.h
+// header. Just enough that we can successfully fail to work there.
+#define MOZ_MINGW_DCOMP_H_INCOMPLETE
+struct IDCompositionColorMatrixEffect : public IDCompositionFilterEffect {};
+struct IDCompositionTableTransferEffect : public IDCompositionFilterEffect {};
+#endif  // defined(__MINGW32__)
+
 namespace mozilla {
 namespace wr {
 
@@ -1820,6 +1806,8 @@
     const color::ColorProfileConversionDesc& conv) {
   auto ret = ColorManagementChain{};
 
+#if !defined(MOZ_MINGW_DCOMP_H_INCOMPLETE)
+
   const auto Append = [&](const RefPtr<IDCompositionFilterEffect>& afterLast) {
     if (ret.last) {
       afterLast->SetInput(0, ret.last, 0);
@@ -1855,6 +1843,9 @@
   ret.dstLinearFromSrcLinear =
       MaybeAppendColorMatrix(color::mat4(conv.dstLinearFromSrcLinear));
   ret.dstTfFromDstLinear = MaybeAppendTableTransfer(conv.dstTfFromDstLinear);
+
+#endif  // !defined(MOZ_MINGW_DCOMP_H_INCOMPLETE)
+
   return ret;
 }
 
